How to build for emscripten?

Hi,

First of all thank you for this nice framework, it has nice cross-platform features.

How do I build it for wasm? in the `build.sh` I saw a branch for emscripten but calling `../tools/build.sh emscripten` it doesn't do the build?

I'm using emsdk `sdk-releases-upstream-89202930a98fe7f9ed59b574469a9471b0bda7dd-64bit` on Linux

Hi,

Yes I did source `emsdk_env.sh` before I ran rainbow's `build.sh`. eg:
```
$ source emsdk/emsdk_env.sh
$ cd repos/rainbow/rainbow-build/
$ ../tools/build/.sh emscripten
----------------------------------------------------------
... snip
-- Build files have been written to: repos/rainbow/rainbow-build
make: *** No targets specified and no makefile found.  Stop.
```

Hi,

I managed to generate cmake by calling emcmake cmake manually. ie: `emcmake cmake -DCMAKE_TOOLCHAIN_FILE=~/emsdk/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake ..` but when i try to build it I hit error:
```
repos/rainbow/src/Audio/cubeb/Mixer.h:15:10: fatal error: 'cubeb/cubeb.h' file not found
#include <cubeb/cubeb.h>
```

Using `../tools/build.sh emscripten` it does generate the `build.ninja` and if I try to build it by running `ninja` I hit the same error:
```
../src/Audio/cubeb/Mixer.h:15:10: fatal error: 'cubeb/cubeb.h' file not found
#include <cubeb/cubeb.h>
```

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -5,7 +5,8 @@ on:
       - master
   pull_request:
 env:
-  WINDOWS_10_SDK: "10.0.19041.0"
+  EMSCRIPTEN_SDK: 2.0.15
+  WINDOWS_10_SDK: 10.0.19041.0
 jobs:
   Windows:
     strategy:
@@ -177,6 +178,29 @@ jobs:
       run: |
         set -o pipefail
         xcodebuild -project build/ios/Rainbow.xcodeproj -scheme Rainbow -arch arm64 CODE_SIGNING_ALLOWED=NO | xcpretty
+  Emscripten:
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v2
+      with:
+        submodules: true
+    - name: Checkout Emscripten SDK
+      uses: actions/checkout@v2
+      with:
+        repository: 'emscripten-core/emsdk'
+        path: '.emsdk'
+    - name: Build
+      run: |
+        set -o pipefail
+        cd .emsdk
+        ./emsdk install $EMSCRIPTEN_SDK
+        ./emsdk activate $EMSCRIPTEN_SDK
+        source ./emsdk_env.sh
+        cd ..
+        mkdir -p build/js
+        cd build/js
+        ../../tools/build.sh emscripten
   Node:
     runs-on: ubuntu-latest
     steps:

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -284,14 +284,24 @@ else()
     src/Audio/AudioFile.h
     src/Audio/Codecs/OggVorbisAudioFile.cpp
     src/Audio/Codecs/OggVorbisAudioFile.h
-    src/Audio/cubeb/Mixer.cpp
-    src/Audio/cubeb/Mixer.h
   )
-  if(APPLE)
+  if(EMSCRIPTEN)
+    add_definitions(-DRAINBOW_AUDIO_AL=1)
     list(APPEND SOURCE_FILES
-      src/Audio/Codecs/AppleAudioFile.cpp
-      src/Audio/Codecs/AppleAudioFile.h
+      src/Audio/AL/Mixer.cpp
+      src/Audio/AL/Mixer.h
     )
+  else()
+    list(APPEND SOURCE_FILES
+      src/Audio/cubeb/Mixer.cpp
+      src/Audio/cubeb/Mixer.h
+    )
+    if(APPLE)
+      list(APPEND SOURCE_FILES
+        src/Audio/Codecs/AppleAudioFile.cpp
+        src/Audio/Codecs/AppleAudioFile.h
+        )
+    endif()
   endif()
 endif()
 
@@ -393,7 +403,7 @@ if(NOT ANDROID)
   find_package(OpenGL REQUIRED)
   if(TARGET OpenGL::OpenGL)
     target_link_libraries(rainbow OpenGL::OpenGL)
-  else()
+  elseif(NOT EMSCRIPTEN)
     target_link_libraries(rainbow ${OPENGL_ext_LIBRARY} ${OPENGL_gl_LIBRARY})
   endif()
 endif()
@@ -419,9 +429,14 @@ elseif(APPLE)
   find_library(CORESERVICES_LIBRARY CoreServices REQUIRED)
   target_link_libraries(rainbow ${CORESERVICES_LIBRARY})
 elseif(UNIX AND NOT MINGW)
-  if(COVERAGE)
-    set(CMAKE_EXE_LINKER_FLAGS "--coverage")
+  if(EMSCRIPTEN)
+    set(CMAKE_EXECUTABLE_SUFFIX ".html")
+    target_link_libraries(rainbow m rt)
+  else()
+    if(COVERAGE)
+      set(CMAKE_EXE_LINKER_FLAGS "--coverage")
+    endif()
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,-z,now -Wl,-z,relro -pie -pthread")
+    target_link_libraries(rainbow m rt stdc++fs)
   endif()
-  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,-z,now -Wl,-z,relro -pie -pthread")
-  target_link_libraries(rainbow m rt stdc++fs)
 endif()

--- a/build/cmake/Audio.cmake
+++ b/build/cmake/Audio.cmake
@@ -1,6 +1,7 @@
 if(EMSCRIPTEN)
   find_package(OpenAL REQUIRED)
   target_link_libraries(rainbow ${OPENAL_LIBRARY})
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s USE_OGG=1 -s USE_VORBIS=1")
 elseif(USE_FMOD_STUDIO)
   set(AUDIO_INCLUDE_DIRS ${LOCAL_LIBRARY}/FMOD/inc)
   if(WIN32)

--- a/build/cmake/FreeType.cmake
+++ b/build/cmake/FreeType.cmake
@@ -1,5 +1,5 @@
 if(EMSCRIPTEN)
-  set(EMCC_FLAGS "${EMCC_FLAGS} -s USE_FREETYPE=1 -s USE_HARFBUZZ=1")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s USE_FREETYPE=1 -s USE_HARFBUZZ=1")
 else()
   # Build FreeType without HarfBuzz to break the circular dependency
   set(FREETYPE_INCLUDE_DIR_freetype2 ${LOCAL_LIBRARY}/FreeType/include)

--- a/build/cmake/SDL2.cmake
+++ b/build/cmake/SDL2.cmake
@@ -6,7 +6,7 @@ if(ANDROID)
   # Pass through.
 elseif(EMSCRIPTEN)
   set(SDL2_INCLUDE_DIRS ${LOCAL_LIBRARY}/SDL/include)
-  set_target_properties(rainbow PROPERTIES LINK_FLAGS "-s DEMANGLE_SUPPORT=1 -s USE_SDL=2")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s DEMANGLE_SUPPORT=1 -s USE_SDL=2")
 elseif(MSVC)
   download_library(
     sdl2-bin

--- a/src/Common/TypeCast.h
+++ b/src/Common/TypeCast.h
@@ -11,13 +11,19 @@
 namespace rainbow
 {
     template <typename T, typename U>
-    constexpr auto narrow_cast(U value)
+    constexpr auto narrow_cast(U value) -> T
     {
-        static_assert(!std::is_same_v<T, U>);
-        static_assert(std::is_signed_v<T> != std::is_signed_v<U> ||
-                      std::numeric_limits<T>::digits <
-                          std::numeric_limits<U>::digits);
-        return static_cast<T>(value);
+        if constexpr (!std::is_same_v<T, U> &&
+                      (std::is_signed_v<T> != std::is_signed_v<U> ||
+                       std::numeric_limits<T>::digits <
+                           std::numeric_limits<U>::digits))
+        {
+            return static_cast<T>(value);
+        }
+        else
+        {
+            return value;
+        }
     }
 
     template <typename T>

--- a/src/Script/NoGame.cpp
+++ b/src/Script/NoGame.cpp
@@ -38,7 +38,8 @@ void NoGame::init_impl(const Vec2i& screen)
 
     render_queue().emplace_back(batch_);
 
-#if !defined(RAINBOW_OS_IOS)  // TODO: Can't access system fonts on iOS
+    // TODO: Can't access system fonts on iOS or in a browser
+#if !defined(RAINBOW_JS) && !defined(RAINBOW_OS_IOS)
     if (error_)
     {
         label_ = std::make_unique<Label>();

--- a/src/Text/SystemFonts.cpp
+++ b/src/Text/SystemFonts.cpp
@@ -28,7 +28,13 @@ namespace
         {
         }
 
-        ~MonospaceFont() { fclose(file_); }
+        ~MonospaceFont()
+        {
+            if (file_ != nullptr)
+            {
+                fclose(file_);
+            }
+        }
 
         [[nodiscard]] auto file_size() const -> size_t
         {
@@ -78,7 +84,7 @@ auto rainbow::text::monospace_font_path() -> czstring
             return path;
     }
 
-    R_ABORT("No monospace fonts found");
+    LOGW("No monospace fonts found");
     return nullptr;
 #elif defined(RAINBOW_OS_WINDOWS)
     return "C:/Windows/Fonts/consola.ttf";

--- a/tools/build.sh
+++ b/tools/build.sh
@@ -13,9 +13,6 @@ fi
 
 function compile {
   case "$1" in
-    "Emscripten")
-      make
-      ;;
     "Ninja")
       $ANALYZER ninja
       ;;
@@ -105,7 +102,7 @@ case $1 in
       exit 1
     fi
     eval run_cmake -DCMAKE_TOOLCHAIN_FILE="$EMSCRIPTEN/cmake/Modules/Platform/Emscripten.cmake" $args &&
-    compile Emscripten
+    compile "$(generator)"
     ;;
   "help")
     $SHELL $0 --help

